### PR TITLE
[dev] Autogenerate OCaml dev files.

### PR DESCRIPTION
**Pull request by ejgallego**
For now we only copy the templates, but we could do more fancy stuff.

This helps to be compatible with build systems that take care of these
files automatically, see:
https://github.com/coq/coq/pull/6857#discussion_r202096579

### CODE REVIEW

**Comment by ejgallego** [comment]
Where to insert the dependency is indeed a good question, suggestions welcome.

File: .gitignore [hunk 1]
```diff
@@ -179,3 +179,8 @@ test-suite/.nra.cache
 
 plugins/ssr/ssrparser.ml
 plugins/ssr/ssrvernac.ml
+
+# ocaml dev files
+.merlin
+ide/.merlin
```
Comment (SkySkimmer) [inline .gitignore]: redundant

File: Makefile.dev [hunk 2]
```diff
@@ -85,13 +85,28 @@ endif
 # But these partial targets could be quite handy for quick builds
 # of specific components of Coq.
 
+###########################################################################
+# OCaml dev files
+###########################################################################
+camldevfiles: $(MERLINFILES) META.coq
+	echo "foo $(MERLINFILES)"
```
Comment (SkySkimmer) [inline Makefile.dev]: ?

File: dev/tools/coqdev.el [hunk 3]
```diff
@@ -33,7 +33,7 @@
 
 (defun coqdev-default-directory ()
   "Return the Coq repository containing `default-directory'."
-  (let ((dir (locate-dominating-file default-directory "META.coq")))
+  (let ((dir (locate-dominating-file default-directory "META.coq.in")))
```
Comment (SkySkimmer) [inline dev/tools/coqdev.el]: If that file is meant to leave it may be worth finding something more robust.

File: Makefile.dev (hunk 2 above)
Comment (ejgallego) [reply Makefile.dev]: Debug code, fixed sorry.

File: dev/tools/coqdev.el (hunk 3 above)
Comment (ejgallego) [reply dev/tools/coqdev.el]: We could use `dune-project` once it is in the tree.

Comment (ejgallego) [reply dev/tools/coqdev.el]: In fact Tuareg should learn to do that automatically.

Comment (SkySkimmer) [reply dev/tools/coqdev.el]: This is coq specific though.

Comment (ejgallego) [reply dev/tools/coqdev.el]: Once we use a standard build system I see little coq-specific stuff there.

**Comment by SkySkimmer** [comment]
little <> nothing

Gaëtan Gilbert

**Comment by ejgallego** [comment]
> little <> nothing

`coqdev.el` can just depend on Tuareg and use its project location.

**Comment by SkySkimmer** [comment]
It's not a submode of tuareg

Gaëtan Gilbert

**Review by gares — APPROVED** [review]

**Review by ppedrot — COMMENTED** [review]
I don't quite know why I am supposed to be an owner here, probably the move of the `ide` file. Absolutely no comment on this PR.

**Review by Zimmi48 — COMMENTED** [review]
I don't quite get how this PR represents an improvement by itself. Is it just a preparatory step in order to have Dune properly generate these files but still allow one to use make and then only copy them? In that case, it means that longer term, these `.in` files would get completely removed, right?

Weak request: could you use this opportunity to move these top-level files (`.merlin.in` and `META.coq.in`) to a sub-folder (for instance `dev/`)?

**Comment by ejgallego** [comment]
> I don't quite get how this PR represents an improvement by itself. Is it just a preparatory step in order to have Dune properly generate these files but still allow one to use make and then only copy them?

Yes. Also this is useful for make maintainers if they want to set version number, flags, etc... automatically.

> Weak request: could you use this opportunity to move these top-level files (.merlin.in and META.coq.in) to a sub-folder (for instance dev/)?

In fact I tried this approach first, but it is worse I think. How would you properly resolve naming and location?

**Comment by Zimmi48** [comment]
> In fact I tried this approach first, but it is worse I think. How would you properly resolve naming and location?

I don't know. Thanks for trying.

**Comment by ejgallego** [comment]
> I don't know. Thanks for trying.

Maybe we can move `META` but the way is now is more standard. Tho, I can confirm that I will remove `META` support from `make` as soon as I can. I do not want to maintain this file. That means in practice when opam packages are built with Dune.

Regarding `.merlin`, the problem is that `dev` folder could also contain a `.merlin` file, so it seems tricky to distinguish. Note that as is stands today, Dune doesn't generate this top-level file.

**Comment by Zimmi48** [comment]
> Note that as is stands today, Dune doesn't generate this top-level file.

But the plan is that it does? Because otherwise, I wouldn't see the point of having this copying step.

**Comment by ejgallego** [comment]
> But the plan is that it does?

It is not needed, instead, it generates a `.merlin` file for each library under the proper directory.

**Comment by Zimmi48** [comment]
Oh I see, so no top-level `.merlin` at all then.

**Comment by ejgallego** [comment]
> Oh I see, so no top-level .merlin at all then.

Yeah, the make system could do the same if they wanted of course.

**Review by maximedenes — APPROVED** [review]
codeowners change looks good

**Comment by ejgallego** [comment]
@gares ping, it seems this is not on your merge queue, however it is approaved.

**Pull request merged**